src/qat/ir/emitter.py

```diff
@@ -0,0 +1,180 @@
+            dependencies = (
+                set([instruction.quantum_targets])
+                if isinstance(instruction.quantum_targets, str)
+                else instruction.quantum_targets
+            )
+        return dependencies
+
+    @property
+    def scheduling_dependencies(self):
+        return self.dependencies
+
+    @property
+    def reliant_instructions(self):
+        return self.reliant
+
+    def __repr__(self):
+        sep = ";" if any(self.reliant) else ""
+        sched_deps = f"[{','.join(self.dependencies)}]" if any(self.dependencies) else ""
+        return (
+            f"{self.instruction.duration}: {sched_deps} {str(self.instruction)}"
+            f"{sep}{';'.join(str(inst) for inst in self.reliant)}"
+        )
+
+
+class QatFile(BaseModel):
+    """In-memory representation of our instruction file."""
+
+    timeline: List[TimelineSegment] = []
+    meta_instructions: InstructionList = InstructionList()
+
+    def add(self, instruction: Instruction, *args, **kwargs):
+        self.timeline.append(TimelineSegment(instruction, *args, **kwargs))
+
+    def add_meta(self, instruction: Instruction):
+        if isinstance(instruction, Return):
+            existing_return = next(
+                iter(meta for meta in self.meta_instructions if isinstance(meta, Return)),
+                None,
+            )
+            if existing_return is not None:
+                existing_return.variables.extend(instruction.variables)
+            else:
+                self.meta_instructions.add(instruction)
+        else:
+            self.meta_instructions.add(instruction)
+
+    @property
+    def instructions(self):
+        return [inst.instruction for inst in self.timeline]
+
+    def get_pp_for_variable(self, target_var):
+        results = []
+        for instruction in self.instructions:
+            if (
+                isinstance(instruction, PostProcessing)
+                and instruction.output_variable == target_var
+            ):
+                results.append(instruction)
+
+        return results
+
+    @property
+    def sweeps(self):
+        return [sw for sw in self.meta_instructions if isinstance(sw, Sweep)]
+
+    @property
+    def repeat(self):
+        return next(
+            iter(sw for sw in self.meta_instructions if isinstance(sw, Repeat)), None
+        )
+
+    @property
+    def return_(self):
+        return next(
+            iter(sw for sw in self.meta_instructions if isinstance(sw, Return)), None
+        )
+
+    def serialize(self):
+        """
+        Serialize the QAT file to a JSON blob.
+        """
+        return self.model_dump_json(indent=4)
+
+    @staticmethod
+    def deserialize(blob):
+        """
+        Takes a JSON blob and returns a QAT file.
+        """
+        blob = from_json(blob)
+        return QatFile(**blob)
+
+
+class InstructionEmitter:
+    """
+    Interface to the LLVM-driven QAT optimization and construction passes. For now we
```

Comment:
imprecise comment, yes it's inspired by LLVM, but it's not driven by llvm ^^, we're just borrowing a few concepts and techniques ^^

Reply:
Haha yeah I agree, this comment was from the legacy code :sweat_smile: